File: ios/Modules/Sources/Federation/api/MirrorDirectory.swift
import Foundation

public protocol MirrorDirectory: Sendable {
    var mirrors: AsyncStream<[Mirror]> { get }
    func refresh() async -> Bool
    func probe(_ mirrorId: String) async -> Bool
}

public struct Mirror: Sendable, Equatable {
    public let id: String
    public let url: String
    public let kind: String              // "tunnel" | "tor" | "ipfs" | "direct"
    public let lastSeenEpoch: Int64?
    public let healthy: Bool
    public init(id: String, url: String, kind: String, lastSeenEpoch: Int64?, healthy: Bool) {
        self.id = id; self.url = url; self.kind = kind
        self.lastSeenEpoch = lastSeenEpoch; self.healthy = healthy
    }
}

File: android/app/src/main/java/sol/vortexx/android/federation/api/MirrorDirectory.kt
package sol.vortexx.android.federation.api

import kotlinx.coroutines.flow.Flow

/**
 * Wave 19 — mirror list probing and federated-room join.
 *
 * When the primary node is unreachable the app falls back through the
 * known mirrors (pulled from `/v1/mirrors` last time we had a session)
 * and, if that fails too, offers multihop-join to any peer reachable
 * from any mirror.
 */
interface MirrorDirectory {
    val mirrors: Flow<List<Mirror>>
    suspend fun refresh(): Boolean
    suspend fun probe(mirrorId: String): Boolean
}

data class Mirror(
    val id: String,
    val url: String,
    val kind: String,          // "tunnel" | "tor" | "ipfs" | "direct"
    val lastSeenEpoch: Long?,
    val healthy: Boolean,
)
